### DISCUSSION TITLE
Escape the box 📦: Godot as an overlay over desktop environment

### DISCUSSION THREAD

**Discussion by caniko** [godotengine/godot-proposals#5388] [Engine Core]
Thinking from a basic principles point of view: Godot and every modern game engine is a window that has a solid background housing amazing art created by individuals and groups.

This design paradigm provides many opportunities, but also closes some doors.

## Request 💌
Get rid of the window; instead, use the desktop environment (DE) as the rendering background. This idea is very similar to Wallpaper Engine on Steam; the difference being the feature is tied to Godot instead of a standalone tool.

### Techinical features
- Mouse and keyboard capture is either optional, or duplicated to the game and desktop environment. In other words, the game should optionally allow the user to interact with desktop environment 
- The game should only cover the desktop environment where it needs to draw assets, meaning I can see the desktop environment.
- Tooling to configure window-size and position both on the developer and enduser side. I expect most games of this genre to cover the entire screen, but in some circumstances it might make sense to cover half, a quarter, or even a custom aspect ratio.
- Continiously updated set of window objects for open windows like the internet browser and file managers. Following methods:
    - Outline
    - Arangement
    - Speed and acceleration
    - Resize
    - Listeners for:
        - is focused/unfocused
        - being expanded/shrunk
        - minimized/maximized
- Windows, Linux, MacOS, and Android support.

## Motivation 🦸
**TLDR** make games that work within the desktop environment!

I see a lot of opportunities for casual gaming inside my own desktop environment. This feature would, therefore, subsume Wallpaper Engine, not contend with it. I am sure translaters from Wallpaper Engine assets to the new Godot tool could even be written.

## Personal 🏃
I have a game idea that I would like to implement some years into the feature, that requires this feature. However, Godot core is outside my field of expertise and someone else would have to make this component. Consider me proactive.

## Final words 🌊
I appreciate any feedback, please correct me if this has no place in Godot (we might disagree, but I am open to arguments). Maybe I missed a detail and this already exists in another open source toolbelt.

Love what Godot stands for!
Cheers!

**Comment by bruvzg** [comment]
> Get rid of the window, and the solid black.

Godot already has support for the windows with per-pixel transparency (in both 3.x and 4.0, availability might vary on GPU driver), so you can create windows without background.

**Reply by caniko** [reply]
Can I also get rid of the window? Is this possible on Android? iOS (would be surprised)?

What you are suggesting is a half measure, I want this to run all the time. Updated the request in this fashion.

**Reply by bruvzg** [reply]
> Is this possible on Android? iOS?

IIRC, transparency is supported on desktop platforms only.

> Can I also get rid of the window?

I not sure what do you mean by `get rid of the window`. You need a window anyway. If you want to draw on top of the desktop (e.g., "clippy" style desktop character), use a topmost window with transparent background (it is supported).

For the "live wallpaper", currently it's not supported (and probably won't work on any platforms except Android and Windows), but it likely would be implemented as a subwindow of the desktop as well.

Would be nice to see some mockup to understand what features you need.

**Reply by caniko** [reply]
Thank you for the feedback! This is great :)

> Would be nice to see some mockup to understand what features you need.

Really good idea, thank you. I added the initial list of features.

> I not sure what do you mean by get rid of the window.

Yes this was bad, I think the feature list clarifies things.

> For the "live wallpaper", currently it's not supported 

Has this been proposed before?

**Reply by bruvzg** [reply]
> Has this been proposed before?

Yes, see https://github.com/godotengine/godot-proposals/issues/965, but it did not get much support. Implementing it is not too hard, and it should be possible to do as a plugin.

Here's a quick POC for Windows and macOS (it's also should be possible to do on Android) - https://github.com/bruvzg/godot/tree/wp_mode

<details>
  <summary>Screen recordings - Click to expand</summary>

https://user-images.githubusercontent.com/7645683/189941032-4bd7f286-a657-4113-80a9-143567756bbc.mp4

https://user-images.githubusercontent.com/7645683/189941047-9036e864-1352-4bd6-9e4b-bc167340f370.mov

</details>

**Reply by caniko** [reply]
I didn't realise it was your doing! Awesome, thank you for starting so fast on the implementation.

How far are you willing to go?

**Reply by caniko** [reply]
Does @Calinou still think it belongs in a third party? Even if it is meant for use in new and creative games? :P

**Reply by Calinou** [reply]
> Does @Calinou still think it belongs in a third party? Even if it is meant for use in new and creative games? :P

There needs to be enough community support for this to be merged in core. In other words, the feature needs to be requested by several people who plan to use it in released projects.

I'm the kind of person who sticks to pure black static wallpapers to save on CPU and battery usage (on OLED displays), so don't count on me to personally support this proposal :slightly_smiling_face:

**Reply by caniko** [reply]
@Calinou, note that this is not WE contender even though that contender might be a gateway to the solution. I want games that are on the DE, small casual games that are continious.

I won't go into the specifics right now, but I have gauged people at work and some friends on the idea of the game: 100% said they are very interested (sample size 8). I'll keep asking people, but note that these are people with very different backgrounds.

So even if you hate WE, which I do too, my wallpaper is also pitch black with the ethereum logo in in black and white in the middle, it might also be something for you too ;)

**Comment by caniko** [comment]
@JoNax97, could you provide some feedback on your thumbdown? It doesn't provide anything of substance other than an empty disapproval, would appreciate a few sentences. Thank you

**Reply by JoNax97** [reply]
I simply don't think this is inside the scope of a project like godot. It's a very specific use case that would require effort and deep knowledge of platform-specific integrations, and this be very costly to develop and maintain, all for little return.

At most, maybe Godot could support borderless windows (if it doesn't yet). But everything else you proposed is too specific imo.

**Reply by caniko** [reply]
I agree that it is niche, but not specific. It is actually a broad genre of games. Morever, I think it gives game developers an exotic alternative to extend the gameplay beyond the black boxes they are designed for today.

Overall it is not very constructive to shut down ideas because of subjective interpretations of scope, but rather better alternatives. If you think there is a project more fitting than Godot please mention it here. Should this idea have a project of its own? Why? I ask because there is a lot, if not complete, overlap on the frontend side.

> deep knowledge of platform-specific integrations, and this be very costly to develop and maintain

This is hypothetical so I will also reply in hypotheticals: Features provide opportunities, opportunities give rise to new ideas and sometimes exciting products.

> At most, maybe Godot could support borderless windows (if it doesn't yet). But everything else you proposed is too specific imo.

Again specific for what? I have a game in mind that requires this feature, specific to that game? What if there is a silent audience, I understand it is a risk, but it should be considered.

**Opinion**
Godot is a general game engine, I think it fits the scope because it allows the artist more creative freedom.

**Reply by JoNax97** [reply]
> I agree that it is niche, but not specific. It is actually a broad genre of games.

Displaying and interacting with games inside the desktop environment It is a very specific feature. regardless of the kind of games you can put in said environment. 

> Overall it is not very constructive to shut down ideas

I'm not shutting down anything. The discussion is open and you're receiving feedback. Part of the proposal system of Godot includes gauging interest/opinion and thumbs up/down are part of that. 

>  If you think there is a project more fitting than Godot please mention it here. Should this idea have a project of its own?

I think this would be best fit a 3rd party godot extension. If, in the development of said extension you find you need access to some lower-level API to build upon, you can always open a proposal for that specifically.

> This is hypothetical so I will also reply in hypotheticals: 

What I said is not hypothetical. It's known that features that depend heavily on OS-specific APIs require people with advanced knowledge in each OS, which are few and far between. 

> Features provide opportunities, opportunities give rise to new ideas and sometimes exciting products.
> What if there is a silent audience, I understand it is a risk, but it should be considered.

Building features upon hypothetical opportunities is contrary to the ethos of the Godot project.

Please don't take this as a personal attack or an attempt to bash your idea. I think it's cool, but ultimately not a good fit to become part of the engine core.

**Reply by caniko** [reply]
> What I said is not hypothetical.

It is hypothetical, you provide no technical reasoning, just opinion that are based on feelings and not technicals-I don't see the value of this kind of input on this thread. In fact it seems that some features already exists according @bruvzg. There are some platform-specific features that have not been implemented.

> I think this would be best fit a 3rd party godot extension

I agree that a subset of my request can become an extension. The underlying transperency features, and DE exposure should probably exist inside core. The window object could probably be an extension, even that I am not so sure, it is so fundemental.

> Please don't take this as a personal attack or an attempt to bash your idea. I think it's cool, but ultimately not a good fit to become part of the engine core.

It is not personal to me, I am fighting for my idea, and you are fighting for yours. We just fundementally disagree, which is fine and good to document, thumbdown just doesn't help anybody. It just creates a smokescreen without answers; thereby, dissatisfaction for those that care (currently just me).

Does Godot have a publicly defined and enforced ethos? Link?

From this point and forward I ask that you use more technical reasoning with respect to the APIs, and their current limitations with respect to challenges on adding them to Godot. Otherwise, thank you for your input!

**Reply by Calinou** [reply]
> Does Godot have a publicly defined and enforced ethos? Link?

[Best practices for engine contributors](https://docs.godotengine.org/en/stable/community/contributing/best_practices_for_engine_contributors.html) is likely the closest.

**Reply by caniko** [reply]
Thank you! I think the problem fits all the criteria, the question is if there is enough community support. :crossed_fingers:

**Comment by EduardoLemos567** [comment]
I didnt read it through and trough...
But just to add, i think we need access of the window transparent background as screen texture, in order to apply effects like blur (acrylic) or any other effect.
Godot should be able to build games, but also building UIs could increase adoption and funds for the engine.

**Comment by ZeromaXHe** [comment]
I think this will be a useful feature for some game ideas, especially for genres like Idler, Simulator, etc.
Nowadays, 2.5 years have passed, and I think more examples show that this proposal is not just for "hypothetical opportunities".
I have seen several Unity games which can run as wallpaper. For example, the **Fantasy Map Simulator** (how to implement that in Godot was also asked by other people in godot forum, [Any way to make a game that runs as a live wallpaper](https://forum.godotengine.org/t/any-way-to-make-a-game-that-runs-as-a-live-wallpaper/105997)), and **Bugtopia**, etc.
I am developing on a game similar to Fantasy Map Simulator, but with Godot engine. Because of Godot lacking this feature, I'm not able to implement this idea in Godot. So I am willing to see this proposal becoming a real feature in Godot someday.

**Reply by paulocoutinhox** [reply]
You solve it?
